rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+      hasLease: true,
+      hasLock: false,
+    },
+    "8d975c5b-cbf6-4418-9afb-7aa3ea34ac90": {
+      version: "7.6.0",
+      lastSeen: [ 1114862, 841295591 ],
+      hasLease: false,
+      hasLock: false,
+    },
+    "3ef25ff1-090a-4335-83a0-307a47712b4e": {
+      version: "7.6.0",
+      lastSeen: [ 1114877, 611368546 ],
+      hasLease: false,
+      hasLock: false,
+    },
+  },
+  oplog: [
+    {op: 'ACQUIRE_LOCK', node: '852bd94e...', timestamp: '2020-04-20T11:58:56.176Z'}
+  ]
+}
+```
+
+### 4.3.2.4 Checking for "weak lease" expiry
+The simplest way to check for lease expiry is to inspect the `lastSeen` value.
+If `lastSeen + expiry_timeout > now` the lock is considered expired. If there
+are clock drift or daylight savings time adjustments, there’s a risk that a
+node loses it’s lease before `expiry_timeout` has occurred. Since losing a
+lock prematurely will not lead to data loss it’s not critical that the
+expiry time is observed under all conditions.
+
+A slightly safer approach is to use a monotonically increasing clock
+(`process.hrtime()`) and relative time to determine expiry. Using a
+monotonically increasing clock guarantees that the clock will always increase
+even if the system time changes due to daylight savings time, NTP clock syncs,
+or manually setting the time. To check for expiry, other nodes poll the
+cluster state document. Once they see that the `lastSeen` value has increased,
+they capture the current hr time `current_hr_time` and starts waiting until
+`process.hrtime() - current_hr_time > expiry_timeout` if at that point
+`lastSeen` hasn’t been updated the lease is considered to have expired. This
+means other nodes can take up to `2*expiry_timeout` to recognize an expired
+lease, but a lease will never expire prematurely. 
+
+Any node that detects an expired lease can release that lease by setting the
+expired node’s `hasLease = false`. It can then attempt to acquire its lease. 
+
+## 4.4 Minimize data loss with mixed Kibana versions
+When multiple versions of Kibana are running at the same time, writes from the
+outdated node can end up either in the outdated Kibana index, the newly
+migrated index, or both. New documents added (and some updates) into the old
+index while a migration is in-progress will be lost. Writes that end up in the
+new index will be in an outdated format. This could cause queries on the data
+to only return a subset of the results which leads to incorrect results or
+silent data loss.
+
+Any in-progress writes such as a bulk operation will continue to run even
+after an index was set to read-only or an alias had been removed. There are
+also certain failure scenarios under which ClusterState (such as aliases or
+index settings) remains inconsistent between nodes for an unbounded
+amount of time. Since Kibana cannot rely on Elasticsearch scripting support
+being enabled, we don't have a mechanism to guarantee that there would be no
```

Comment:
> Since Kibana cannot rely on Elasticsearch scripting support being enabled

How significantly better could we make this if we could require painless to be enabled on the `.kibana` index? If that is an option that gives us leverage, we should discuss making that a requirement in 8.0

Reply:
I should maybe remove or reword this since it suggests that scripting could fix this problem. We could use scripting to add additional protection but it still won't guarantee no data loss.

With scripts enabled we could turn all updates into scripted updates that only update a document if the migration version hasn't been incremented. Whereas the write alias strategy for 8.x cannot prevent existing bulk operations from succeeding, this would provide a guarantee even for in-flight bulk updates. However, we cannot do the same for create operations. If we continuously polled and migrated outdated documents we could migrate created outdated documents, but there would still be a period of time where a query on the index would return inconsistent results.

At the end of the day it feels like if we can't tell users "your data is 100% safe with mixed kibana versions" then the amount of effort we spend on plugging the gaps has much less value.

Reply:
> At the end of the day it feels like if we can't tell users "your data is 100% safe with mixed kibana versions" then the amount of effort we spend on plugging the gaps has much less value.

100% agree, there are diminishing returns on this. Basically I see a lot of value in not having any problems 95% of the time, and even more value in 100% of the time, but not much of a difference between 95% and 99%. So I agree if we can't guarantee it, then it's most likely not worth the effort.